### PR TITLE
Replace Vuex with requestData object that uses Pinia

### PR DESCRIPTION
**Pull request by matthew-white**

### CODE REVIEW

**Comment by matthew-white** [comment]
I'm going to rebase on `master` in order to resolve conflicts. When I do so, I'll squash most of the commits in this PR.

@ktuite and I have reviewed all of src/ and most of test/. The main changes that we still need to review have to do with test utility functions.

**Review by matthew-white — COMMENTED** [review]
Notes from interactive code review

**Review by matthew-white — COMMENTED** [review]
Notes from interactive code review

**Comment by matthew-white** [comment]
I'm going to rebase onto the `master` branch again in order to resolve conflicts with #636. The current commit is 83e386a3b65490a9f23f562540a4f6e6d7945e62, and the last files to review are:

- test/components/navbar.spec.js
- test/components/submission/decrypt.spec.js
- test/request-data/resource.spec.js
- test/util/http.js
- test/util/router.js

File: test/util/http.js [hunk 1]
```diff
@@ -569,61 +570,68 @@ class MockHttp {
...
+    const removeHandler = router != null
+      ? router.onError(error => { this._errorFromRouter = error; })
+      : noop;
+
+    try {
+      if (this._location != null) await router.push(this._location);
+      if (this._mount != null) this._component = this._mount();
+      const mountTriggeredNavigation = this._mount != null && router != null &&
+        router.currentRoute.value === START_LOCATION;
+      if (mountTriggeredNavigation) await router.isReady();
```
Comment (matthew-white) [inline test/util/http.js:614]: Instead:

```js
if (router != null) await router.isReady();
```

\+ add a comment about how `this._mount()` can trigger navigation.

File: test/util/http.js [hunk 2]
```diff
@@ -569,61 +570,68 @@ class MockHttp {
...
+
+      if (this._request != null) {
+        // If there has been a navigation, then wait for any async components
+        // associated with the route to load.
+        if (this._location != null || mountTriggeredNavigation) await wait();
```
Comment (matthew-white) [inline test/util/http.js:619]: Can we simplify the check here as well?

**Review by matthew-white — COMMENTED** [review]
Notes from interactive code review

**Review by ktuite — APPROVED** [review]
Woohoo! Glad to get this reviewed and into the main branch!!

File: test/util/request-data.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            const { createResource } = useRequestData();
+            createResource(composableOrName);
+          }));
+      // Using a strategy similar to withSetup().
+      const setup = () => {
+        for (const composable of composables)
+          composable();
+        return noop;
+      };
+      vtuMount({ setup }, {
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply test/util/request-data.js:76]: This is needed in order to avoid a cyclical dependency: test/util/lifecycle.js imports test/util/request-data.js by way of test/util/container.js, so test/util/request-data.js can't import `mount()` from test/util/lifecycle.js. I'll add a comment about this.

File: test/components/navbar.spec.js [hunk 4]
```diff
@@ -1,27 +1,28 @@
...
-      return load('/login', {
-        container: { router: testRouter(addHook) },
-        attachTo: document.body
-      })
+      let called = false;
+      return load('/login')
+        .beforeAnyResponse(app => {
+          app.vm.$router.currentRoute.value.should.equal(START_LOCATION);
+          app.vm.$router.afterEach(() => {
+            // TODO. See what happens if the assertion fails.
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply test/components/navbar.spec.js]: I've removed TODOs except for one related to comments and one in vue.config.js, both of which I'll come back to soon.

File: src/components/form/show.vue [hunk 5]
```diff
@@ -122,55 +102,53 @@ export default {
...
                   url: draftUrl,
-                  update: ['enketoId'],
+                  patch: ({ data }) => {
+                    if (this.formDraft.isDefined())
+                      this.formDraft.get().enketoId = data.enketoId;
+                  },
                   alert: false
-                }]);
+                });
+                if (this.formDraft.isEmpty()) return true;
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply src/components/form/show.vue]: I checked that if a `patch` callback throws an error, that won't break Frontend. I also wrote some related tests, for example, verifying that even if a `patch` callback throws an error, the state of the resource is cleaned up.

I remembered that I think I've actually encountered this case where `this.formDraft.isEmpty()` within the `patch` callback — so we probably should make some change here and not just revert. That said, I think the underlying cause is different from what I thought. It's not a race condition between two requests: a GET request for the draft and its Enketo ID; and another request that will delete the draft (either publishing or abandoning it). I still think that that's an edge case. Instead, I think it's more likely that the sequence I observed is that the GET request for the draft is sent _after_ the response to the request that deletes the draft. The GET request results in a 404.1 Problem, but that's processed as an expected response to the request, not an error, so the `patch` callback is called when it shouldn't be. I'm going to make two changes related to this:

1. Check before sending the GET request whether the draft has been deleted. That should be enough to handle this particular sequence. This can be seen in 2d09dcbc249f5e74156901a606bb38d551b62df0. (I also moved around some checks, which reduces duplication and I think is ultimately clearer.)
2. More broadly, this points to an issue with `Option` resources in this PR. On the `master` branch right now, the main request for `formDraft` specifies a `fulfillProblem` option that will accept a 404.1 Problem; but the request within the `callWait()` callback does not specify `fulfillProblem`. However, in this PR, I've set up `Option` resources so that they always specify `fulfillProblem`. I could override the `request()` method for `Option` resources so that they don't specify that default `fulfillProblem` option when a `patch` callback is specified. But that would be more complexity than I'm looking to add in this area and I think would bring up inheritance questions around resources, which I want to avoid. I'm going to go back to how things are on the `master` branch here: `Option` resources will not automatically specify `fulfillProblem`. That also means that I'll be able to remove the `requestDefaults` property. This change can be seen in c1e336e090e54a2f4f0a5547f59647c8a5112c08.

I think this might be an interesting area for further thought. For example, maybe we should automatically cancel a request for a resource if the resource's `data` property is set during the request? I'm going to keep some of the logic and the test for the edge case in case that's a useful example to consider as we continue thinking about this area.

Despite this long comment, all this doesn't amount to a huge change and will mostly be returning to how things already work on the `master` branch. Given that, I don't think that @ktuite and I necessarily need to review these commits. However, @ktuite, please ping me if you want to discuss any of this!
